Firestore/core/src/firebase/firestore/remote/stream.mm

```diff
@@ -228,41 +223,64 @@
 
 void Stream::Stop() {
   EnsureOnQueue();
-
   LOG_DEBUG("%s stop", GetDebugDescription());
 
-  if (!IsStarted()) {
+  Close(Status::OK());
+}
+
+void Stream::Close(const Status& status) {
+  // This function ensures that both graceful stop and stop due to error go
+  // through the same sequence of steps. While it leads to more conditional
+  // logic, the benefit is reducing the chance of divergence across the two
+  // cases.
+
+  EnsureOnQueue();
+  bool graceful_stop = status.ok();
+
+  // Step 1 (both): check current state.
+  if (graceful_stop && !IsStarted()) {
+    // Graceful stop is idempotent.
     return;
   }
+  HARD_ASSERT(IsStarted(), "Trying to close a non-started stream");
 
-  // If the stream is in the auth stage, gRPC stream might not have been created
-  // yet.
-  if (grpc_stream_) {
+  // Step 2 (both): cancel any outstanding timers (they're guaranteed not to
+  // execute).
+  CancelIdleCheck();
+  backoff_.Cancel();
+
+  // Step 3 (both): increment close count, which invalidates the auth callback,
```

Comment:
nit: The other platforms invalidate more than the auth callback using this method. Consider rephrasing to something like "... which invalidates long-lived callbacks, guaranteeing they won't execute against a new instance of the stream."

Reply:
Done.